File: fileMonitor.py
# -*- coding: utf-8 -*-
"""
-------------------------------------------------
   File Name：     fileMonitor
   Description :
   Author :       CoolCat
   date：          2019/1/3
-------------------------------------------------
   Change Activity:
                   2019/1/3:
-------------------------------------------------
"""
__author__ = 'CoolCat'

from watchdog.observers import Observer
from watchdog.events import *
import time

global doWithout
global showDir


class FileEventHandler(FileSystemEventHandler):


    def __init__(self):
        FileSystemEventHandler.__init__(self)


    def on_moved(self, event):

        ###  文件移动显示青色
        if doWithout in event.src_path:
            pass
        elif event.is_directory and showDir == "y":
            print(time.strftime('[%H:%M:%S]:') + "\033[0;36m%s\033[0m" % "directory moved from {0} to {1}".format(event.src_path,event.dest_path))
            pass
        elif event.is_directory == False:
            print(time.strftime('[%H:%M:%S]:') + "\033[0;36m%s\033[0m" % "file moved from {0} to {1}".format(event.src_path,event.dest_path))

    def on_created(self, event):

        ###  文件创建显示绿色

        if doWithout in event.src_path:
            pass
        elif event.is_directory and showDir == "y":
            print(time.strftime('[%H:%M:%S]:') + "\033[5;32m%s\033[0m" % "directory created:{0}".format(event.src_path))
            pass
        elif event.is_directory == False:
            print(time.strftime('[%H:%M:%S]:') + "\033[5;32m%s\033[0m" % "file created:{0}".format(event.src_path))

    def on_deleted(self, event):

        ###  文件删除显示红色
        if doWithout in event.src_path:
            pass
        elif event.is_directory and showDir == "y":
            print(time.strftime('[%H:%M:%S]:') + "\033[0;31m%s\033[0m" % "directory deleted:{0}".format(event.src_path))
            pass
        elif event.is_directory == False:
            print(time.strftime('[%H:%M:%S]:') + "\033[0;31m%s\033[0m" % "file deleted:{0}".format(event.src_path))

    def on_modified(self, event):

        ###  文件修改显示蓝色
        if doWithout in event.src_path:
            pass
        elif event.is_directory and showDir == "y":
            print(time.strftime('[%H:%M:%S]:') + "\033[0;34m%s\033[0m" % "directory modified:{0}".format(event.src_path))
            pass
        elif event.is_directory == False:
            print(time.strftime('[%H:%M:%S]:') + "\033[0;34m%s\033[0m" % "file modified:{0}".format(event.src_path))


def help():
    print(""" 
    _____________
   < FileMonitor >
    -------------
      /\_)o<
     |       | 
     | O . O |
      \_____/
    By CoolCat
      """)

if __name__ == "__main__":
    help()

    import sys
    if sys.version_info.major == 2:
        try:
            monitorDir = raw_input(time.strftime('[%H:%M:%S]:') + "Please enter a directory:")
            doWithout = raw_input(time.strftime('[%H:%M:%S]:') + "Unnecessary directory:")
            showDir = raw_input(time.strftime('[%H:%M:%S]:') + "Display directory changes(y or n):")
        except:
            pass
    else:
        monitorDir = input(time.strftime('[%H:%M:%S]:') + "Please enter a directory:")
        doWithout = input(time.strftime('[%H:%M:%S]:') + "Unnecessary directory:")
        showDir = input(time.strftime('[%H:%M:%S]:') + "Display directory changes(y or n):")

    print(time.strftime('[%H:%M:%S]:') + "\033[0;31m%s\033[0m" % "FileMonitor is running...")

    # # monitorDir = "/Users/CoolCat/php"
    # print(type(monitorDir))


    observer = Observer()
    event_handler = FileEventHandler()
    observer.schedule(event_handler,monitorDir,True)
    observer.start()
    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        observer.stop()
    observer.join()

File: fileMonitor.php
<?php
/*
-------------------------------------------------
   File Name：     fileMonitor
   Description :
   Author :       CoolCat
   date：          2019/1/3
-------------------------------------------------
   Change Activity:
                   2019/1/3:
-------------------------------------------------
*/
class FileMonitor
{
	private $dir;
	private $i=0;
	private $files = [];
	private $filesize = [];
	public function __construct($argv)
	{
		$this->dir = $this->getparam($argv);
		$this->getfiles($this->dir);
		print "[+] total:".count($this->files[$this->i])."\n\r";
		$this->i++;
		while(true)
		{
			$this->getfiles($this->dir);
			if(isset($this->files[$this->i-1]) && ((count($this->files[$this->i])>count($this->files[$this->i-1]))))
			{
				print "[+] total:".count($this->files[$this->i])."\n\r";
				print "[*] addfile: ".implode('|',array_diff($this->files[$this->i],$this->files[$this->i-1]))."\n\r";
			}
			if(isset($this->files[$this->i-1]) && ((count($this->files[$this->i])<count($this->files[$this->i-1]))))
			{
				print "[+] total:".count($this->files[$this->i])."\n\r";
				print "[*] deletefile: ".implode('|',array_diff($this->files[$this->i-1],$this->files[$this->i]))."\n\r";
			}
			if(isset($this->filesize[$this->i-1]))
			{
				array_map(function($v,$val,$key){
					if($v != $val)
					{
						print "[*] updatefile:{$key}\n\r";
					}
				},$this->filesize[$this->i-1],$this->filesize[$this->i],array_keys($this->filesize[$this->i]));
			}
			$this->i++;
			if($this->i>=30)
			{
				$this->files = [];
				$this->filesize = [];
				$this->i = 0;
			}
		}
	}

	private function getparam($argv)
	{
		foreach($argv as $key=>$val)
		{
			if($val == "--dir")
			{
				return is_dir($argv[$key+1])?$argv[$key+1]:exit("[-] directory does not exist!");
			}
		}
	}

	private function getfiles($dir)
	{
		if(is_dir($dir))
		{
			$d = scandir($dir);
			foreach($d as $v)
			{
				if($v != '.' && $v != '..')
				{
					if(is_dir("{$dir}/{$v}"))
					{
						$this->getfiles("{$dir}/{$v}");	
					}
					else
					{
						$this->files[$this->i][] = "{$dir}/{$v}";
						$this->filesize[$this->i]["{$dir}/{$v}"] = filesize("{$dir}/{$v}");
					}
				}
			}
		}
		else
		{
			$this->files[$this->i][] = $dir;
			$this->filesize[$this->i][$dir] = filesize($dir);
		}
	}
}
print " 
    _____________
   < FileMonitor >
    -------------
      /\_)o<
     |       | 
     | O . O |
      \_____/
    By CoolCat
";
new FileMonitor($argv);
?>